components/sup/src/manager/file_watcher.rs

```diff
@@ -1248,2217 +327,1778 @@ impl<C: Callbacks, W: Watcher> FileWatcher<C, W> {
+            | ModifyKind::Metadata(MetadataKind::WriteTime) => { // v5.0.0: poll (file was written)
+                self.handle_file_modification(notify_event);
+            }
+            ModifyKind::Name(rename_mode) => {
+                self.handle_rename_event(*rename_mode, notify_event)
+            }
+            ModifyKind::Metadata(MetadataKind::Any) // see comments below
+            | ModifyKind::Metadata(MetadataKind::Ownership) // see comments below
+            | ModifyKind::Metadata(MetadataKind::Permissions) // see comments below
+            | ModifyKind::Metadata(MetadataKind::AccessTime) // we don't care about mere file access
+            | ModifyKind::Metadata(MetadataKind::Extended) // FSEvents: macOS Extended Attributes
+            | ModifyKind::Metadata(MetadataKind::Other) // FSEvents: Finder info
+            | ModifyKind::Other => { // v5.0.0 - Not Implemented 
+                self.log_ignored_event(notify_event)
+                // Changes to ownership and permissions can make files and directories appear or
+                // disappear. However, in v5.0.0 Ownership is only used for FSEvents and Permissions
+                // isn't implemented by anything only being found in a single test.
+                //
+                // This section also account for MetadataKind::Any because in notify v5.0.0 all uses
+                // of MetadataKind::Any seem to be related to metadata/attrs broadly but in an way 
+                // that we can't know if its ownership, permissions, or something else without going
+                // deeper than we probably want to.
+                // - FSEVents: StreamFlags::INODE_META_MOD 'only described as "metadata changed"'
+                // - inotify: EventMask::ATTRIB which might be permissions and ownership
+                // - kqueue: kqueue::Vnode::Attrib might also be permissions and ownership
             }
         }
-        debug!("event in dirs: {:?}", dirs);
-        self.callbacks.event_in_directories(&dirs);
     }
 
-    fn handle_process_path(&mut self) -> Result<Vec<PathsAction>> {
-        let mut actions = Vec::new();
-        match self.paths.process_path_or_defer_if_unsettled() {
-            None => (),
-            Some(directories) => {
-                for directory in directories {
-                    self.watcher
-                        .watch(&directory, RecursiveMode::NonRecursive)?;
-                }
-                if let Some(ref path) = self.paths.real_file {
-                    actions.push(PathsAction::NotifyFileAppeared(path.clone()));
-                }
+    fn handle_remove_event(&mut self, remove_kind: RemoveKind, notify_event: &Event) {
+        trace!("FileWatcher::handle_remove_event - {:?}", remove_kind);
+        match remove_kind {
+            RemoveKind::Any // v5.0.0: FSEvent, poll, and windows
+            | RemoveKind::File => { // v5.0.0: FSEvent and inotify
+                self.handle_file_disappearance(notify_event);
```

Comment:
The exact events depend on what backend is in use and what is being watched.  You can look at [remove-events.txt](https://github.com/habitat-sh/habitat/files/10514070/remove-events.txt) to see some different combinations for the poll watcher and inotify backends.  That said, I don't think there is an issue because while multiple events could be generated they should be effectively filtered our by the fact that we are trigger only when the path of the watched file is matched.  So if we were watching a directory and received multiple events only a file removal matching the path of our file should trigger an event.

Reply:
I've just renamed and cleaned up a test and I'm about to add two more. In a bit I'll do a commit with delete_directory_enclosing_watched_file(), rename_directory_enclosing_watched_file(), and move_directory_enclosing_watched_file().  I suspect those along with their *_poll_watcher counterparts will surface any issues.  I'll message you when I've pushed these changes to the repo.

Reply:
Correction, there will be two methods.  The delete test is cleaned up and implemented and the move test should be soon.  However the rename isn't going to work because as per the doc "On Unix, if from is a directory, to must also be an (empty) directory. If from is not a directory, to must also be not a directory. In contrast, on Windows, from can be anything, but to must not be a directory." so there is no joy in rename land for directories.